src/google/protobuf/compiler/java/java_message.cc

```diff
@@ -124,7 +124,7 @@ void ImmutableMessageGenerator::GenerateStaticVariables(io::Printer* printer) {
 
     // The descriptor for this type.
     printer->Print(vars,
-      "$private$static final com.google.protobuf.Descriptors.Descriptor\n"
+      "$private$static com.google.protobuf.Descriptors.Descriptor\n"
```

Comment:
This can't be final any more because it isn't set from inside the `static {}` block directly. If your static block calls a static method, and that method sets then variable, then even if the method is `private static` and not called from anywhere else, the variable it sets can't be `static`.

If this is an issue, I may have a way to work around this. Will probably take until the weekend at least though.